conduit.go

```diff
@@ -0,0 +1,325 @@
+		password: {{$service.credentials.password}}
+		db: {{$service.credentials.name}}
+	{{end}}
+{{end}}
+
+Press Ctrl+C to shutdown.
+`
+
+func init() {
+	ConnectService.Flags().Int64VarP(&LocalPort, "local-port", "p", 7080, "start selecting local ports from")
+}
+
+func waitForConnection(port int64) chan error {
+	timeout := 3 * time.Second
+	connection := make(chan error)
+	go func() {
+		defer close(connection)
+		tries := 0
+		for {
+			if tries > 5 {
+				time.Sleep(2 * time.Second)
+			} else {
+				time.Sleep(1 * time.Second)
+			}
+			tries++
+			conn, err := net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", port), timeout)
+			if err != nil {
+				if tries < 15 {
+					continue
+				}
+				connection <- fmt.Errorf("timeout waiting for tunnel to start: %s", err)
+				break
+			}
+			defer conn.Close()
+			connection <- nil
+			break
+		}
+	}()
+	return connection
+}
+
+var ConnectService = &cobra.Command{
+	Use: "conduit [flags] SERVICE_INSTANCE [-- COMMAND]",
+	Example: `  Create a tunnel between your machine and a remote running service:
+  cf conduit my-service
+  
+  Run a local application that uses VCAP_SERVICES environment to connect to remote services:
+  cf conduit my-service-1 my-service-2 -- /path/to/app
+    
+  Export a postgres database:
+  cf conduit postgres-instance -- pg_dump -f backup.sql
+
+  Import a postgres script:
+  cf conduit postgres-instance -- psql < backup.sql
+
+  Dump a mysql database:
+  cf conduit mysql-instance -- mysqldump --all-databases > backup.sql
+
+  Import a mysql script:
+  cf conduit mysql-instance -- psql < backup.psql
+  `,
+	Short: "enables temporarily binding services to local running processes",
+	Long:  "spawns a temporary application, binds your desired service and creates an ssh tunnel from the application to your local machine enabling communication directly with the remote service.",
+	Args: func(cmd *cobra.Command, args []string) error {
+		if cmd.ArgsLenAtDash() > -1 {
+			if cmd.ArgsLenAtDash() < 1 {
+				return errors.New("requires at least one SERVICE_INSTANCE argument to be specified")
+			}
+		} else if len(args) < 1 {
+			return errors.New("requires at least one SERVICE_INSTANCE argument to be specified")
+		}
+		return nil
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		lg := func(args ...interface{}) {
+			if Verbose {
+				fmt.Fprintln(Stderr, args...)
+			}
+		}
+		var serviceInstanceNames []string
+		var runargs []string
+		if cmd.ArgsLenAtDash() > -1 {
+			serviceInstanceNames = args[:cmd.ArgsLenAtDash()]
+			runargs = args[cmd.ArgsLenAtDash():]
+		} else {
+			serviceInstanceNames = args
+			runargs = []string{}
+		}
+		cf := Cf{Cli: conn, Verbose: Verbose}
+
+		tempAppDir, err := cf.generateEmptyApp()
+		defer os.RemoveAll(tempAppDir)
+		if err != nil {
+			return err
+		}
+
+		appName := filepath.Base(tempAppDir)
+		lg("Deploying tunnel", appName)
+		err = cf.pushAppWithoutRoute(appName, tempAppDir)
+		defer func() {
```

Comment:
Again, this `defer` should come after the error checking.

Reply:
It's possible for pushes to "fail" but still leave app data that needs cleaning up and I have seen this manifest thanks to the office timeout errors.

So as it stands this defer is in the correct place.... however the pushApp func should really be responsible for guaranteeing no sideeffects on errors! so I'll work on cleaning that up.